Dumping files

Quaker762 opened the pull request:
Added a function to dump virtual files to disk, so as to aid in reverse
engineering any file types that are currently unknown.

z33ky committed:
Remove arch.hpp and math.hpp

Unused. Vector extensions have taken their intended place.

z33ky committed:
Return from main() instead of calling exit()

See issue #53.

Quaker762 committed:
Initial implementation of Virtual Files

Title explains it all. These are supposed to somewhat mimic C file IO,
as some file types will need to be seeked (Mesh files, animation files
etc). I think this makes a nice way to handle Virtual Files from arc
sections.

However, I feel the std::vector buffer may be a waste of space, and
there could be a MUCH better way to do this without buffering the entire
file (a HUGE waste of memory).

z33ky committed:
Add assert.hpp

This commit adds some ASSERT* macros.
Behavior is controlled through macros set by the build system.

This commit also removes "-masm=intel" to make scottt's debugbreak work.

z33ky committed:
Generalize sh3_arc_file::read_error as error<result_enum>

z33ky committed:
Use error for sh3_texure::load_result.

z33ky committed:
Code simplification of sh3_arc_file::read_error::message()

Quaker762 committed:
Fixed assert problem in 'vfile.cpp' and Modified ReadData()

Totally forgot that the assert error is thrown if 0 is returned from the
expression.... Ooops!

Also modified ReadData() to set a partial read AFTER we work out how
many bytes we want to read.

z33ky committed:
Fix sh3_graphics::texture::Load signature in implementation

This error was introduced in a1f9f378edd9b13ecc43d3109823e02ddf8ae982.

z33ky committed:
Add SH3/graphics/texture.cpp to CMakeLists.txt

z33ky committed:
Add vfile.cpp to CMakeLists.txt and fix compile error

z33ky committed:
Fix clang warnings

vfile::fpos became std::size_t. It also makes sense that we're able to
address the whole range of fsize, which is also std::size_t.

sh3_arc_vfile had two logic errors that were fixed. One is an invalid if
conditional that was unintendedly constant, and one if conditional
checked against the wrong variable.

z33ky committed:
Use clamping instead of jump-to-0 for past-the-end sh3_arc_vfile::Seek

If a file is sought past the end, no data should be returned on
ReadData().

Quaker762 committed:
Updated codeblocks project

Updated the codeblocks project (it was HORRIBLY out of date and didn't
have correct compiler warnings enabled).

Quaker762 committed:
Updated .gitignore

Updated .gitignore so as to ignore the stupid codeblocks *.depend file
(none relative paths, it is generated on build anyway)

Quaker762 committed:
Removed -masm=intel flag

z33ky commented on source/SH3/arc/vfile.cpp:
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
+    std::stringstream ss(fname);
+    std::string item;
+
+    while(getline(ss, item, '/')); // Throw out all strings except the last one (our file name)
```
It'd be cheaper to look for the last `/` in the string (see `std::string::rfind`) and passing an iterator pair to `std::ofstream`'s constructor (`begin(fname) + pos` and `end(fname)`).

z33ky commented on source/SH3/arc/vfile.cpp:
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
+
+    std::ofstream out_file(item);
+    for(int i = 0; i < buffer.size(); i++)
+        out_file << buffer[i];
```
The "streaming operators" (`<<`, `>>`) are meant for formatted data. We're dealing with binary data, so use `std::ofstream::write` instead. And you don't need to do it item by item then, you can just `write(buffer.data(), buffer.size())`.

z33ky commented on source/SH3/arc/vfile.cpp:
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
+
+    while(getline(ss, item, '/')); // Throw out all strings except the last one (our file name)
+
+    std::ofstream out_file(item);
```
Check if opening fails.

z33ky commented on source/SH3/arc/vfile.cpp:
```diff
@@ -100,3 +102,23 @@ std::size_t sh3_arc_vfile::ReadData(void* destination, std::size_t len, read_err
+
+void sh3_arc_vfile::Dump2Disk()
+{
+    if(!open || buffer.empty())
```
Shouldn't be just create an empty file if `buffer.empty()`? Is this an error?

Quaker762 replied:
Not so much an error, it just doesn't dump anything to the disk to save a few clock cycles. I didn't really see the point of creating an empty file, though I did initially have it in.

Quaker762 commented:
@z33ky  I think I've just noticed a problem with how I have this set up. Some mesages may have more than one attribute, but if we change `currentAttr` during `Load()`, the old attribute will be lost. How do you think we should store all the attributes??

z33ky commented:
`currentAttr` is part of the other pull request #65. I'll answer there.
Could you add `msg.cpp` to the `source/CMakeLists.txt` as well?

Quaker762 commented:
Oops, I didn't even notice that. Is this PR ready for a merge?

> Could you add msg.cpp to the source/CMakeLists.txt as well?

I have in the last pushed commit in #67

z33ky commented:
Doesn't look like it
```
$ git diff --name-only upstream/master quaker/dump
SILENT HILL 3 Redux.cbp
include/SH3/arc/vfile.hpp
include/SH3/system/msg/msg.hpp
source/SH3/arc/section.cpp
source/SH3/arc/types.cpp
source/SH3/arc/vfile.cpp
source/SH3/system/msg/msg.cpp
```

Quaker762 commented:
Oops, I didn't realise you meant to the dump branch, I'll do that and rebasr asap.

z33ky commented:
Well that is confusing. It seems you have some changes intended for #65 in here, as was also previously the case as seen by my `git diff` output before your rebase. You should remove those commits (including of course adding `msg.cpp` to `CMakeLists.txt`) from this pull request, especially since the message stuff is still subject to change.

Quaker762 commented:
Okay, I really fucked that rebase up badly (I must've pushed a button accidentally...). I'm working on fixing it now...

Quaker762 committed:
Dumping files

Added a function to dump virtual files to disk, so as to aid in reverse
engineering any file types that are currently unknown

Quaker762 commented:
I'm deleting this branch and closing, I've fucked this up too hard.
